### ISSUE TITLE
v0.36.x: wire live CUDA probe_fn into pick_batch_size from SFTTrainerWrapper

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#64]
## Background

v0.36.0 (Part D) shipped the OOM-probe auto batch-size infrastructure: ``utils/batch_probe.py`` (``probe_batch_size`` try-halve-then-double loop, cache key, atomic-rename JSON cache at ``~/.soup/batch_cache.json`` with 0600 perms and env-override containment), plus ``training.auto_batch_size_strategy: auto|static|probe``.

The trainer-side wiring deliberately passes ``probe_fn=None`` to ``pick_batch_size``: the cache short-circuit + static-estimate fallback ship now, but the live forward+backward+step probe against the real model is not invoked. From ``soup_cli/trainer/sft.py``:

```python
batch_size = pick_batch_size(
    ...
    probe_fn=None,  # CUDA probe wired in v0.36.x patch — for
                    # now we honour the cache + static estimate
                    # so the surface ships with no regression.
)
```

This means first-run users on a new ``(model, max_length, quant, lora_r, gpu)`` tuple still get the static-formula estimate (which under-counts activations / gradient buffers / optimizer state). Repeat runs hit the cache and short-circuit, but only after at least one OOM-then-retry training run has populated it.

## Proposed fix path

1. In ``SFTTrainerWrapper`` (after model + LoRA are loaded but before ``train_ds``), build a ``probe_fn`` closure:
   - Construct a tiny synthetic batch (``input_ids`` of shape ``(B, max_length)`` — random ints under vocab).
   - Run ``model(...).loss.backward()`` followed by ``optimizer.step() + zero_grad()`` once.
   - Wrap with ``torch.cuda.empty_cache()`` + ``torch.cuda.reset_peak_memory_stats()`` between probes.
   - Return ``True`` on success; let ``torch.cuda.OutOfMemoryError`` propagate (already in default ``oom_exceptions``).
2. After probing, ``del`` the probe optimizer state and call ``empty_cache`` again.
3. Pass the ``probe_fn`` to ``pick_batch_size`` only on CUDA (``self.device == "cuda"``).
4. Add an integration test under a ``@pytest.mark.gpu`` marker that mocks ``torch.cuda.OutOfMemoryError`` at certain sizes and asserts the probe converges + caches.

## Acceptance criteria

- [ ] ``probe_fn=None`` is gone from the SFT path on CUDA runs.
- [ ] Cache file written after the first probe-run on a fresh ``(model, ...)`` key.
- [ ] CPU-only runs still short-circuit to static estimate without warnings (unless ``strategy="probe"`` is explicit, in which case the existing v0.36.0 review-fix warning fires).
- [ ] Tests cover the convergence path (mocked OOM + happy path) and the cache short-circuit on second invocation.
- [ ] The code-review HIGH about silent fallback (#code-review-medium-13 in v0.36.0 review notes) is fully closed.

## Files

- ``soup_cli/trainer/sft.py`` (the ``probe_fn=None`` site)
- ``tests/test_batch_probe.py`` (add CUDA-mocked integration test)

**Comment by snakefood3232** [comment]
The missing "probe_fn" in `pick_batch_size`, causing a static-formula estimate usage, caught my attention. I'd handle this by implementing a `probe_fn` closure in `SFTTrainerWrapper`, executing synthetic batch runs to dynamically determine batch size, following your proposed steps closely. Can have a PR up in 2 days. I've built similar batch size tuning systems with live CUDA probes before.

**Comment by MakazhanAlpamys** [comment]
**Note for contributors:** the previous claim on this issue was from an automated spam account that posted 10+ formulaic "I'd handle this by..." comments across the repo without ever submitting a PR, ignored two direct warnings, and has now been blocked.

This issue is **open and unclaimed** -- any genuine contributor is welcome to pick it up.

That said: this is internal roadmap work touching many files. If you'd like a friendlier first contribution, see #11, #12, #17, #26, #36, #50, #51, or #55.

**Comment by MakazhanAlpamys** [comment]
Shipped in v0.40.3. See [release notes](https://github.com/MakazhanAlpamys/Soup/releases/tag/v0.40.3).

Try: `soup train --config soup.yaml` with `auto_batch_size_strategy: probe` in `training:` — the SFT trainer now runs ONE forward+backward+step on a synthetic batch per candidate before training. Result is cached per `(model, max_length, quant, lora_r, gpu)` tuple. Non-SFT trainer expansion is tracked separately.
